File: java/src/offertwelve/offertwelve.java
package offertwelve;

public class offertwelve{
    private static boolean[][] visited;
    public static boolean exist(char[][] board, String word){
        int m = board.length, n = board[0].length;
        visited = new boolean[m][n];
        char[] words = word.toCharArray();
        boolean res = false;
        int cur = 0;
        for (int i = 0; i<m; i++){
            for (int j = 0; j<n;j++){
                res = helper(board, words, cur, i, j);
                if (res == true) return true;
            }
        }
        return false;
    }

    public static boolean helper(char[][] board, char[] word, int cur, int i, int j){
        if (cur == word.length) return true;
        if (i<0 || i>board.length-1 || j<0 || j>board[0].length-1 || visited[i][j] || board[i][j] != word[cur]) return false;
        visited[i][j] = true;
        cur++;
        boolean res = helper(board, word, cur, i+1, j) || 
                    helper(board, word, cur, i-1, j) || 
                    helper(board, word, cur, i, j+1) ||
                    helper(board, word, cur, i, j-1);
        visited[i][j] = false;
        return res;
    }
    public static void main(String[] args) {
        char[][] board = new char[][] {{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
        String word = "ABCCED";
        boolean res = exist(board, word);
        System.out.println(res);
    }
}

File: go/src/offertwelve/offertwelve.go
package main

import "fmt"

func exist(board [][]byte, word string) bool {
	if len(board) == 0 {
		return false
	}
	isVisited := make([][]bool, len(board))
	for i := 0; i<len(board); i++ {
		isVisited[i] = make([]bool, len(board[0]))
	}
	for i := 0; i<len(board); i++ {
		for j :=0; j<len(board[0]); j++{
			if board[i][j] == word[0] {
				if bfs(board, word, 0, i, j, isVisited){
					return true
				}
			}
		}
	}
	return false
}

func bfs(board [][]byte, word string, cur int, i int, j int, isVisited [][]bool) bool {
	if cur == len(word) {
		return true
	}
	if i < 0 || j < 0 || i == len(board) || j == len(board[0]) || isVisited[i][j] || board[i][j] != word[cur] {
		return false
	}
	isVisited[i][j] = true
	res := bfs(board, word, cur+1, i+1, j, isVisited) ||
		bfs(board, word, cur+1, i, j+1, isVisited) ||
		bfs(board, word, cur+1, i-1, j, isVisited) ||
		bfs(board, word, cur+1, i, j-1, isVisited)
	isVisited[i][j] = false
	return res
}

func main() {
	board := [][]byte{{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}}
	word := "ABCCED"
	res := exist(board, word)
	fmt.Println(res)
}